src/Data/Bifunctor/Biap.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DeriveTraversable          #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+
+#include "bifunctors-common.h"
+#ifndef MIN_VERSION_semigroups
+#define MIN_VERSION_semigroups(x,y,z) 0
+#endif
+
+-----------------------------------------------------------------------------
+-- |
+-- Copyright   :  (C) 2008-2016 Edward Kmett
+-- License     :  BSD-style (see the file LICENSE)
+--
+-- Maintainer  :  Edward Kmett <ekmett@gmail.com>
+-- Stability   :  provisional
+-- Portability :  portable
+--
+----------------------------------------------------------------------------
+module Data.Bifunctor.Biap 
+ ( Biap(..)
+ ) where
+
+import GHC.Generics
+#if __GLASGOW_HASKELL__ < 710
+import Data.Foldable
+import Data.Monoid
+import Data.Traversable
+#endif
+import Data.Biapplicative
+import Data.Bifoldable
+import Data.Coerce
+#if LIFTED_FUNCTOR_CLASSES
+import Data.Functor.Classes
+#endif
+#if MIN_VERSION_base(4,9,0) || MIN_VERSION_semigroups(0,16,2)
+import Data.Semigroup
+#endif
+import Control.Monad.Fail (MonadFail)
+import GHC.Base hiding (Any)
+
+-- | Pointwise lifting of a class over two arguments, using
+-- 'Biapplicative'.
+--
+-- Classes that can be lifted include 'Monoid', 'Num' and
+-- 'Bounded'. Each method of those classes can be defined as lifting
+-- themselves over each argument of 'Biapplicative'.
+--
+-- @
+-- mempty        = bipure mempty          mempty
+-- minBound      = bipure minBound        minBound
+-- maxBound      = bipure maxBound        maxBound
+-- fromInteger n = bipure (fromInteger n) (fromInteger n)
+--
+-- negate = bimap negate negate
+-- 
+-- (+)  = biliftA2 (+)  (+)
+-- (<>) = biliftA2 (<>) (<>)
+-- @
+--
+-- 'Biap' is to 'Biapplicative' as 'Data.Monoid.Ap' is to
+-- 'Applicative'.
+--
+-- 'Biap' can be used with @DerivingVia@ to derive a numeric instance
+-- for pairs:
+--
+-- @
+-- newtype Numpair a = Np (a, a)
+--  deriving (Semigroup, Monoid, Num, Bounded)
+--  via Biap (,) a a
+-- @
+-- 
+newtype Biap bi a b = Biap { getBiap :: bi a b }
+ deriving ( Eq
+          , Ord
+          , Show
+          , Read
+          , Enum
+          , Functor
+          , Foldable
+          , Traversable
+          , Alternative
+          , Applicative
+          , Generic
+          , Generic1
+          , Monad
+          , MonadFail
+          , MonadPlus
+#if LIFTED_FUNCTOR_CLASSES
+          , Eq1
+          , Ord1
+#endif LIFTED_FUNCTOR_CLASSES
```

Comment:
Don't put anything after `#endif`; preprocessors often choke on these (see [here](https://travis-ci.org/ekmett/bifunctors/jobs/633885995#L602), for example).